pkg/cache/v3/linear.go

```diff
@@ -312,90 +370,72 @@ func (cache *LinearCache) GetResources() map[string]types.Resource {
+	//  - we use the version as some form of hash or resources known, and we can then consider it as a way to correctly verify whether all resources are unchanged.
+	// For now it is not done as:
+	//  - for the first case, while the protocol documentation does not explicitly mention the case, it does not mark it impossible and explicitly references unsubscribing from wildcard
+	//  - for the second one we could likely do it with little difficulty if need be
+	watch := ResponseWatch{Request: request, Response: value, subscription: sub}
 
 	cache.mu.Lock()
 	defer cache.mu.Unlock()
 
-	switch {
-	case err != nil:
-		stale = true
-		staleResources = request.GetResourceNames()
-		cache.log.Debugf("Watch is stale as version failed to parse %s", err.Error())
-	case len(request.GetResourceNames()) == 0:
-		stale = (lastVersion != cache.version)
-		if stale {
-			cache.log.Debugf("Watch is stale as cache version %d differs for wildcard watch %d", cache.version, lastVersion)
-		}
-	default:
-		for _, name := range request.GetResourceNames() {
-			// When a resource is removed, its version defaults 0 and it is not considered stale.
-			if lastVersion < cache.versionVector[name] {
-				stale = true
-				staleResources = append(staleResources, name)
-			}
-		}
-		if stale {
-			cache.log.Debugf("Watch is stale with stale resources %v", staleResources)
-		}
-	}
-	if stale {
-		cache.respond(watch, staleResources)
+	response := cache.computeSotwResponse(watch)
+	if response != nil {
+		watch.Response <- response
 		return nil, nil
 	}
+
+	watchId := cache.nextSotwWatchID()
 	// Create open watches since versions are up to date.
-	if len(request.GetResourceNames()) == 0 {
+	if sub.IsWildcard() {
 		cache.log.Infof("[linear cache] open watch for %s all resources, system version %q", cache.typeURL, cache.getVersion())
-		cache.watchAll[watch] = struct{}{}
+		cache.watchAll[watchId] = watch
 		return func() {
 			cache.mu.Lock()
 			defer cache.mu.Unlock()
-			delete(cache.watchAll, watch)
+			delete(cache.watchAll, watchId)
 		}, nil
 	}
 
-	cache.log.Infof("[linear cache] open watch for %s resources %v, system version %q", cache.typeURL, request.ResourceNames, cache.getVersion())
-	for _, name := range request.GetResourceNames() {
+	cache.log.Infof("[linear cache] open watch for %s resources %v, system version %q", cache.typeURL, sub.SubscribedResources(), cache.getVersion())
+	for name := range sub.SubscribedResources() {
 		set, exists := cache.watches[name]
 		if !exists {
-			set = make(watches)
+			set = make(map[int64]ResponseWatch)
 			cache.watches[name] = set
 		}
-		set[watch] = struct{}{}
+		set[watchId] = watch
 	}
 	return func() {
 		cache.mu.Lock()
 		defer cache.mu.Unlock()
-		cache.removeWatch(watch)
+		cache.removeWatch(watchId, watch.subscription)
 	}, nil
 }
 
+func (cache *LinearCache) nextSotwWatchID() int64 {
+	return atomic.AddInt64(&cache.sotwWatchCount, 1)
```

Comment:
I don't think it is likely to happen, but perhaps it should panic on overflow, for good measure?

Reply:
Added a check

Reply:
Although the code as is probably works, it has 2 memory barriers which smell racy.
I think the following would be much more clear:

```
next := atomic.AddInt64(&cache.sotwWatchCount, 1)
if next < 0 {
	panic("watch id count overflow")
}
return next
```

Reply:
Good remark, updated